### ISSUE TITLE
Need Support for Async implimentation for SOAP calls.

### ISSUE THREAD

**Issue by sreekar12** [quarkiverse/quarkus-cxf#636]
SOAP calls are not accepting as Uni<Object> as return type. Could you please help us any approach to make SOAP calls as Async.

For now i was using return type as Java class instead of Uni. For converting Uni to Java class i was using join() OR get().

**Note:** Join() and get() blocks until it gives response.

Could you please suggest any approach instead of above note....

**Comment by shumonsharif** [comment]
@sreekar12 Not sure if it may be helpful, but have a look at the code posted here:  https://github.com/quarkiverse/quarkus-cxf/issues/4#issuecomment-1106487964

**Comment by shumonsharif** [comment]
Hi @sreekar12 - Your question is unrelated to the extension, but I believe you may be able to find some answers on SO, for example [here](https://stackoverflow.com/a/69347025).

**Comment by tbkahuna48** [comment]
The code is attempting to make several asynchronous parallel backend calls and aggregate the results. The current cxf implementation seems to require a join on the completion stages which is a blocking implementation. So the question is , how can the calls be made without blocking. sreekar referred to Uni because most quarkus extensions work with Unis. The question though is how can they be made in parallel in a non-blocking way.

**Comment by linktech1** [comment]
Here's another description of the problem. The Quarkus CXF extension does not currently support asynchronous response types in the service interface and need to convert async types (Uni, Completion or Future) to standard response types. This conversion results in the async type to block until backside responses are returned.

**Comment by linktech1** [comment]
I tried using the @UseAsyncMethod annotation, https://cxf.apache.org/javadoc/latest-3.2.x/org/apache/cxf/annotations/UseAsyncMethod.html

The async methods I have implemented per the link above are not called. If the async methods were called I would also expect them to be called on the event-loop-thread

============================= Code ===============================
@WebService(endpointInterface = "io.quarkiverse.cxf.it.server.GreetingWebService", serviceName = "GreetingWebService")
@BindingType(javax.xml.ws.soap.SOAPBinding.SOAP12HTTP_BINDING)
public class GreetingWebServiceImpl implements GreetingWebService {

    private static final Logger LOG = LoggerFactory.getLogger(GreetingWebServiceImpl.class);

    @Inject
    HelloBean helloResource;

    @Override
    @UseAsyncMethod
    public String reply(@WebParam(name = "text") String text) {
        LOG.info("Calling reply ");
        return helloResource.getHello() + text;
    }

    @Override
    @UseAsyncMethod
    public String ping(@WebParam(name = "text") String text) throws GreetingException {
        if (text.equals("error")) {
            throw new GreetingException("foo", "bar");
        }
        LOG.info("Calling reply ");
        return helloResource.getHello() + text;
    }

    public Future<?> replyAsync(@WebParam(name = "text") String text,
            final AsyncHandler<GreetMeSometimeResponse> asyncHandler) {
        LOG.info("Calling replyAsync ");

        final ServerAsyncResponse<GreetMeSometimeResponse> response = new ServerAsyncResponse<GreetMeSometimeResponse>();
        GreetMeSometimeResponse tempResponse = new GreetMeSometimeResponse("Hello " + text);
        tempResponse.setResponseType(text);

        response.set(tempResponse);
        return response;
    }

    public Future<?> pingAsync(@WebParam(name = "text") String text,
            final AsyncHandler<GreetMeSometimeResponse> asyncHandler) throws GreetingException {
        LOG.info("Calling pingAsync ");
        if (text.equals("error")) {
            throw new GreetingException("foo", "bar");
        }
        final ServerAsyncResponse<GreetMeSometimeResponse> response = new ServerAsyncResponse<GreetMeSometimeResponse>();
        GreetMeSometimeResponse tempResponse = new GreetMeSometimeResponse("Hello " + text);
        tempResponse.setResponseType(text);
        response.set(tempResponse);
        return response;
    }

}

**Comment by shumonsharif** [comment]
> @shumonshariff Sorry, some additional context may help. The code is attempting to make several asynchronous parallel backend calls and aggregate the results. The current cxf SOAP implementation seems to require a join on the completion stages which is a blocking implementation. So the question is , how can the calls be made without blocking. Sreekar referred to Uni because most quarkus extensions work with Unis, the question though is how can the SOAP calls be made in parallel in a non-blocking way. This is specific to CXF SOAP as it is , among the many other implementations of parallel asynchronous calls using quarkus extension libraries, the only we as yet aren't able to figure out how to make in a non-blocking manner. [bryan.g.mclane@kp.org](mailto:bryan.g.mclane@kp.org)

@tbkahuna48 Did you or Sreekar look at the example provided in my first comment?  That example demonstrates exactly what you are describing?

**Comment by shumonsharif** [comment]
> I tried using the @UseAsyncMethod annotation, https://cxf.apache.org/javadoc/latest-3.2.x/org/apache/cxf/annotations/UseAsyncMethod.html
> 
> The async methods I have implemented per the link above are not called. If the async methods were called I would also expect them to be called on the event-loop-thread

@linktech1 The `@UseAsyncMethod` as I understand is for CXF service providers, not consumers.  Are you all trying to consume a web service (that's the impression I get from all the comments by @sreekar12 and @tbkahuna48) or provide a CXF service for others to consume?

**Comment by linktech1** [comment]
Yes. We are providing a service that calls other services. We have the the backend service clients being called asynchronously (they return Futures) but we also want the frontside service implementation to also be non-blocking. Looking for the same support as the Quarkus Rest services where the service is completely non-blocking and runs entirely on the vertex-event-loop.

**Comment by linktech1** [comment]
Anything on Quarkus CXF supporting service provider interfaces with non-blocking async behavior that runs on the vertex-io-thread?

**Comment by shumonsharif** [comment]
@linktech1 I haven't had much time to look into this, and likely won't any time soon ... My take is that this is going to be a fairly major effort, and I don't believe the JAX-WS spec will allow us to mimic the same features / patterns as the Quarkus REST services.  We may at most be able to possibly integrate the `@UseAsyncMethod` functionality offered in CXF.

@ppalaga @famod Any thoughts on this issue?

**Comment by linktech1** [comment]
I think the @UseAsyncMethod feature would work if the event-loop-thread is used. Async should be disabled if Quarkus detects other blocking code.

**Comment by ppalaga** [comment]
Thanks for reaching out, @sreekar12, @linktech1 and @tbkahuna48. 

Would adding a proper support for `@UseAsyncMethod` solve your needs?

**Comment by linktech1** [comment]
The requirement is to enable Quarkus reactive capability and not spin up any additional worker threads (run on the event loop) if the operation flow is non blocking. It looks like the current implementation creates a new Thread for the ServerAsyncResponse in this example: https://cxf.apache.org/javadoc/latest/org/apache/cxf/annotations/UseAsyncMethod.html

We would like to see a Uni or CompletionStage be used to be consistent with Quarkus reactive return types.

Cheers,

Keith Link
***@***.***

**Comment by dabkhazi** [comment]
Sorry if my question seems stupid, but for an asynchronous client is it necessary to modify the server part of the service by adding copies of methods that return the Future?

**Comment by dabkhazi** [comment]
As a result, I created a test project to check the functionality of the asynchronous client as described in the HTTP Async Transport section. I created a server and a client for testing. Here is the link to the project on GitHub:
https://github.com/dabkhazi/quarkus-soap-async-test
However, I am unable to use the asynchronous client due to the following error:

java.lang.ClassCastException: class org.apache.cxf.jaxws.JaxwsResponseCallback cannot be cast to class io.netty.util.concurrent.Future 
(org.apache.cxf.jaxws.JaxwsResponseCallback and io.netty.util.concurrent.Future are in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader)

The test project uses Java 17 and Quarkus 3.15.1. Do you have any ideas on what might be going wrong?
